src/operator/adaptive_avg_pooling.cc

```diff
@@ -0,0 +1,230 @@
+
+        real grad_delta = gradOutput_p_d[oh*osizeW +ow] / kH / kW;
+
+        int ih, iw;
+        for (ih = istartH; ih < iendH; ih++) {
+          for (iw = istartW; iw < iendW; iw++) {
+            /* update gradient */
+            gradInput_p_d[ih*isizeW + iw] += grad_delta;
+          }
+        }
+      }
+    }
+  }
+}
+
+
+template<typename xpu, typename DType, typename AccReal>
+void AdaptiveAvgPoolUpdateOutput(mshadow::Stream<cpu> *s,
+                                           const std::vector<TBlob> &input,
+                                           const std::vector<TBlob> &output) {
+  Tensor<xpu, 4, DType> itensor = input[0].get<xpu, 4, DType>(s);
+  Tensor<xpu, 4, DType> otensor = output[0].get<xpu, 4, DType>(s);
+
+  DType *input_data = itensor.dptr_;
+  DType *output_data = otensor.dptr_;
+
+  int64_t sizeB  = itensor.size(0);
+  int64_t sizeD  = itensor.size(1);
+  int64_t isizeH = itensor.size(2);
+  int64_t isizeW = itensor.size(3);
+
+  int64_t istrideB = get_stride<xpu, 4, DType>(itensor, 0);
+  int64_t istrideD = get_stride<xpu, 4, DType>(itensor, 1);
+  int64_t istrideH = get_stride<xpu, 4, DType>(itensor, 2);
+  int64_t istrideW = get_stride<xpu, 4, DType>(itensor, 3);
+
+  int64_t osizeH = otensor.size(2);
+  int64_t osizeW = otensor.size(3);
+
+  int64_t b;
+#pragma omp parallel for private(b) \
+num_threads(engine::OpenMP::Get()->GetRecommendedOMPThreadCount())
+  for (b = 0; b < sizeB; b++) {
+    SpatialAdaptiveAveragePooling_updateOutput_frame<DType>(
+      input_data+b*istrideB, output_data+b*sizeD*osizeH*osizeW,
+      sizeD,
+      isizeH, isizeW,
+      osizeH, osizeW,
+      istrideD,
+      istrideH, istrideW);
+  }
+}
+
+
+template<typename xpu, typename DType, typename AccReal>
+void AdaptiveAvgPoolUpdateGradInput(mshadow::Stream<cpu> *s,
+                                              const std::vector<TBlob> &input,
+                                              const std::vector<TBlob> &output) {
+  Tensor<xpu, 4, DType> gradOut = input[0].get<xpu, 4, DType>(s);
+  Tensor<xpu, 4, DType> gradIn = output[0].get<xpu, 4, DType>(s);
+
+  DType *gradOutput_data = gradOut.dptr_;
+  DType *gradInput_data = gradIn.dptr_;
+
+  int64_t sizeB  = gradIn.size(0);
+  int64_t sizeD  = gradIn.size(1);
+  int64_t isizeH = gradIn.size(2);
+  int64_t isizeW = gradIn.size(3);
+
+  int64_t osizeH = gradOut.size(2);
+  int64_t osizeW = gradOut.size(3);
+
+  int64_t b;
+#pragma omp parallel for private(b) \
+num_threads(engine::OpenMP::Get()->GetRecommendedOMPThreadCount())
+  for (b = 0; b < sizeB; b++) {
+    SpatialAdaptiveAveragePooling_updateGradInput_frame<DType>(
+      gradInput_data+b*sizeD*isizeH*isizeW, gradOutput_data+b*sizeD*osizeH*osizeW,
+      sizeD,
+      isizeH, isizeW,
+      osizeH, osizeW);
+  }
+}
+
+
+DMLC_REGISTER_PARAMETER(AdaptiveAvgPoolParam);
+
+NNVM_REGISTER_OP(AdaptiveAvgPool2D)
+.describe(R"code(
+Applies a 2D adaptive average pooling over an input signal composed of several input planes.
```

Comment:
what does adaptive pooling do?
why is the output H x W? Isn't it the same with input size?
what does the output_size parameter do?

Please add more documentation. See documentation of other operators like Pooling. Also see the original pytorch doc.

Reply:
The doc is adapted from http://pytorch.org/docs/0.3.1/nn.html?highlight=adaptive#torch.nn.AdaptiveAvgPool2d